### ISSUE TITLE
reentrant checkpoint cause cyclic reference if run_fn capture object that reference output

### ISSUE THREAD

**Issue by YouJiacheng** [pytorch/pytorch#84413]
```python
import gc

import torch
from torch.utils.checkpoint import checkpoint

def foo(x) -> torch.Tensor:
    data = {'dummy': x * 0} # add dummy data to show "leak" caused by cyclic reference
    def bar(x):
        data # let run_function capture data
        return x
    # x = bar(x) # no checkpoint ok
    x = checkpoint(bar, x, use_reentrant=True) # non-reentrant implementation ok: ctx.run_function is the cause
    data['x'] = x # create cyclic reference
    return x

x = torch.zeros(4096, 4096).cuda().requires_grad_(True)
for _ in range(50):
    xx = foo(x).sum()
    xx.backward()
    # gc.collect() # with gc.collect is ok - not real memory leak, only a cyclic reference
    print(torch.cuda.max_memory_allocated())
```
Should we add `del ctx.run_function` in `backward` of `CheckpointFunction`(still leak if no `backward` is called) or dynamically create `CheckpointFunction` capturing `run_function` or just let it go?
I also experience a (possibly) real memory leak which cannot solved by `gc.collect` in a project, but I cannot figure out a MWE. (This problem was solved by adding `del ctx.run_function`)

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by soulitzer** [comment]
The behavior in the repro provided is mostly expected.
- We may not necessarily want to `del ctx.run_function` because someone could potentially run backward a second time.
- It seems like `gc.collect()` works because there is indeed a cycle that can be detected from python.
- It may not work in cases where (1) there is still something referring to the cycle. In this case you'll want to `del xx`, or (2) part of that cycle is not traversable by the python gc.
- Another way to refer to the ctx from outside of the checkpoint implementation is just to look at the grad_fn, so `del xx.grad_fn.run_function` may also work here. 

As you have mentioned, other workarounds here are probably to use use_reentrant=True, and to avoid capturing tensors refering to the graph in functions that will be saved on the graph.

**Comment by albanD** [comment]
If gc.collect works, then it is not a memory leak.
The only thing we can try to do is to use weakref if posisble in this case to avoid the cycle?

**Comment by YouJiacheng** [comment]
@soulitzer Thanks for your speedy and comprehensive reply!
- Yes, the potential double backward blocks `del ctx.run_function`. I thought about this before I open this issue.
- Can we say that real memory leak will never happen if `CheckpointFunction` works as expected(i.e. if there is no bug) and `run_fn` itself won't leak? And, run `gc.collect` might have non negligible performance implication, which is undesirable.
- Hmmm, that is really hard to debug.
- I know for the first time that `xx.grad_fn is ctx`, thanks! But this way is impractical in common usecases: we only checkpoint part of model, and it is not easy to access the intermediate result after `loss.backward()`.

@albanD Thanks for help! I considered weakref. But I haven't had any idea where we can use weakref properly.

**Comment by soulitzer** [comment]
The following is an example where gc.collect() will not help (and so there is an actual leak):
```
import gc

import torch
from torch.utils.checkpoint import checkpoint

def foo(x) -> torch.Tensor:
    data = {'dummy': x * 0} # add dummy data to show "leak" caused by cyclic reference
    def bar(x):
        data # let run_function capture data
        return x
    # x = bar(x) # no checkpoint ok
    x = checkpoint(bar, x, use_reentrant=True) # non-reentrant implementation ok: ctx.run_function is the cause
    y = x.clone() # <-- extra clone added
    data['x'] = y # create cyclic reference
    return y

x = torch.zeros(4096, 4096).cuda().requires_grad_(True)
for _ in range(50):
    xx = foo(x).sum()
    xx.backward()
    # gc.collect() # with gc.collect is ok - not real memory leak, only a cyclic reference
    print(torch.cuda.max_memory_allocated())
```
That is because the cycle is now:
- grad_fn (checkpoint) -> run_function -> closure -> captured x -> grad_fn (clone) -> grad_fn (checkpoint)
Previously it was
- grad_fn (checkpoint) -> run_function -> closure -> captured x -> grad_fn (checkpoint)

This happens because pyobject traversal only travels one-hop into the backward graph past the tensor, so appending that extra node to the graph prevents gc from clearing the cycle.

**Comment by YouJiacheng** [comment]
Hmmm, IIRC cpython gc will recursively traversal pyobject, is it not the case?🤔

**Comment by soulitzer** [comment]
Yes that is the case, but the backward graph is actually in c++. We observe the above because there is a special logic added in the python binding code to check only one step into this c++ graph

**Comment by soulitzer** [comment]
Closing because there should be no bug in checkpointing logic
